### PR TITLE
Rustup to *rustc 1.10.0-nightly (764ef92ae 2016-05-19)* and bump to 0.0.69

### LINKED ISSUE #945: Cannot install clippy
Hello.
I recently switched to using `rustup` so I decided I would try the `clippy` sub-command but I am unable to install `clippy`.

Here are the steps I used:

```bash
~ rustup default nightly
info: using existing install for 'nightly-x86_64-unknown-linux-gnu'
info: default toolchain set to 'nightly-x86_64-unknown-linux-gnu'

  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.10.0-nightly (9c6904ca1 2016-05-18)

~ cargo -V
cargo 0.11.0-nightly (fb03dad 2016-05-18)

~ cargo install clippy > result
```
(see [attached result.txt file](https://github.com/Manishearth/rust-clippy/files/273619/result.txt))

I use ArchLinux and I installed `rustup` using this command:

```bash
yaourt -S rustup
```

Is this a clippy or a rustup issue?

How can I install `clippy` using rustup?

Thanks for your help.

### LINKED ISSUE #946: Build fail 0.0.68
https://travis-ci.org/crabtw/rust-bindgen/jobs/132243299
Any idea where it comes from?

### PR DESCRIPTION
**Pull request by mcarton**
Ref https://github.com/rust-lang/rust/pull/33654.
Fix #945. Fix #946.

### CODE REVIEW

**Comment by mcarton** [comment]
Don’t know why the tests are failing :sweat:

**Comment by oli-obk** [comment]
Some are failing, because there's no way to detect that a for loop was expanded, except to check the HIR for blocks that state that they are for loop expansions. But the HirMap::parent method doesn't give the actual parent if run on anything inside the expansion, but it gives the parent of the for loop.

**Comment by oli-obk** [comment]
ahaa, this probably occurrs because the idents in HIR don't carry information about expansions anymore, but are just names now: https://github.com/rust-lang/rust/pull/33654

**Comment by oli-obk** [comment]
more info: https://github.com/rust-lang/rust/pull/30145#issue-119779170

> Besides name resolution `unhygienic_name` is used in some lints and debuginfo. `unhygienic_name` can be very well approximated by "reverse renaming" `token::intern(name.as_str())` or even plain string `name.as_str()`, except that it would break gensyms like `iter` in desugared for loops. This approximation is likely good enough for lints and debuginfo, but not for name resolution, unfortunately (see #27639), so `unhygienic_name` has to be kept.

**Comment by oli-obk** [comment]
probably [this](https://github.com/rust-lang/rust/blob/master/src/librustc/hir/lowering.rs#L1530) should be `gensym("_result")` instead

**Comment by oli-obk** [comment]
or even better, it should simply not create a local variable at all

**Comment by mcarton** [comment]
> ahaa, this probably occurrs because the idents in HIR don't carry information about expansions anymore, but are just names now: rust-lang/rust#33654

I know, I linked that in the PR :smile:
I meant “I don’t know why `name.unhygienize()` seems different from `ident.unhygienic_name` here”.

> probably [this](https://github.com/rust-lang/rust/blob/master/src/librustc/hir/lowering.rs#L1530) should be `gensym("_result")` instead
> or even better, it should simply not create a local variable at all

That local variable is weird, but I can’t update *rustc* :sweat_smile: 

We might need to rewrite the lint to find that out otherwise but we can’t make it an early-lint because there seem to be no way to get a `syntax::ast` expression’s parent.

@Manishearth ?

**Comment by oli-obk** [comment]
> I know, I linked that in the PR :smile:

whoops :)

> That local variable is weird

it's there due to some lifetime issues with putting values into the tail expression

> I meant “I don’t know why name.unhygienize() seems different from ident.unhygienic_name here”.

Well... `unhygienize` uses the thread local interner, which is [cleared](https://github.com/rust-lang/rust/blob/master/src/librustc_driver/driver.rs#L261) several times before HIR lints run, IF there is a previous compilation error [1](https://github.com/rust-lang/rust/blob/master/src/librustc_driver/driver.rs#L985), [2](https://github.com/rust-lang/rust/blob/master/src/librustc_driver/driver.rs#L917)

Easy fix: don't run lints if there were previous errors

**Comment by Manishearth** [comment]
> That local variable is weird, but I can’t update rustc 

sure you can :smile:

**Comment by Manishearth** [comment]
We could switch over UsedUnderscoreBinding to use a regular macro check and switch the shadow pass to ignore macros too.

**Comment by oli-obk** [comment]
> We could switch over UsedUnderscoreBinding to use a regular macro check and switch the shadow pass to ignore macros too.

temporarily? because if permanent, we loose the checks if the name is used in a macro like `println!("{}", _something);`

**Comment by Manishearth** [comment]
I'm not sure if we do?

**Comment by oli-obk** [comment]
> I'm not sure if we do?

Also the issue with the for loop won't go away, since it's not from a macro expansion, but from ast -> hir lowering, which isn't recorded.

**Comment by mcarton** [comment]
With `in_macro` we lose `println!("{}", _foo);` and `_result` in for loop desugaring still warns.

File: src/shadow.rs [hunk 1]
```diff
@@ -66,7 +66,7 @@ fn check_fn(cx: &LateContext, decl: &FnDecl, block: &Block) {
     let mut bindings = Vec::new();
     for arg in &decl.inputs {
         if let PatKind::Ident(_, ident, _) = arg.pat.node {
-            bindings.push((ident.node.unhygienic_name, ident.span))
```
Comment (petrochenkov) [inline src/shadow.rs]: Why isn't `unhygienize()` used in this file?

Comment (mcarton) [inline src/shadow.rs]: I does now, but the lint still has problems.

File: src/shadow.rs [hunk 2]
```diff
@@ -335,8 +335,8 @@ struct ContainsSelf {
 }
 
 impl<'v> Visitor<'v> for ContainsSelf {
-    fn visit_ident(&mut self, _: Span, ident: Ident) {
-        if self.name == ident.unhygienic_name {
+    fn visit_name(&mut self, _: Span, name: Name) {
+        if self.name == name {
```
Comment (petrochenkov) [inline src/shadow.rs]: One more missing `unhygienize()` here.

Comment (petrochenkov) [inline src/shadow.rs]: If some lints don't work due to gensyms after proper replacement of all `unhygienic_name`s with `unhygienize()` , they need to be moved to AST.

Comment (Manishearth) [inline src/shadow.rs]: Fair point.

Comment (mcarton) [inline src/shadow.rs]: > One more missing unhygienize() here.

Stupid me. That lint seems ok now. Thanks.

> they need to be moved to AST

I would have done that but `used_underscore_binding` needs to get the parent of an expression. With `hir` that’s in `rustc::hir::map::Map`, there does not seem to be an `ast` equivalent. I’ll see if we can rewrite the lint to not use that.

**Comment by mcarton** [comment]
Another day, another problem:
```rust
    with_vec.unwrap_or(vec![]);
    //~^ERROR use of `unwrap_or`
    //~|HELP try this
    //~|SUGGESTION with_vec.unwrap_or_else(|| vec![]);
```

is now reported as:
```rust
tests/compile-fail/methods.rs:        with_vec.unwrap_or_else(|| < [ _ ] > :: into_vec ( box [ $ ( $ x ) , * ] ));
```

This looks to be because of https://github.com/rust-lang/rust/pull/33730 or https://github.com/rust-lang/rust/pull/33712. Looks like `snippet` won’t work with macros anymore, we could get the right span with `CodeMap::source_callsite`  guess, but how would we know when we have to use it?

**Comment by Manishearth** [comment]
@jseyfried @nrc Tips on what to use in place of snippet here?

**Comment by Manishearth** [comment]
Am I right in assuming that the broken snippets are the only thing left to fix? In that case, we could temporarily patch the tests and fix this later.

**Comment by mcarton** [comment]
> Am I right in assuming that the broken snippets are the only thing left to fix?

No, `used_underscore_binding` still warns on loop desugaring `_result` variable.

**Comment by Manishearth** [comment]
We can quick patch that to ignore variables with that name too.

**Comment by mcarton** [comment]
After 3 rustups the tests are finally passing \o/
I made `used_underscore_binding` allow-by-default and ignore `_result` for now. Also I commented out a macro-span test.
@Manishearth are you ok with that? Can we publish?

**Comment by Manishearth** [comment]
Yep, but file follow-ups for the corners we cut. I'll poke at it later.

**Comment by jseyfried** [comment]
@Manishearth 
> Tips on what to use in place of snippet here?

This was caused by [#33712](https://github.com/rust-lang/rust/pull/33712), which fixed the span of expanded expressions that are the root of their expansion. These expressions used to have the span of their expansion's invocation (the `call_site`).

Note that expanded items have always had the correct span, for example:
```rust
macro_rules! m {
    () => { const fn f() {} } //< This is the span of the expanded item, ...
}

m!(); //< ... not this.
```
The most direct way to fix the fallout here is to compute what the span would have been the change:
```rust
fn expansion_root_to_call_site(self: &CodeMap, mut span: Span) -> Span {
    self.with_expn_info(span.expn_id, |info| info.map(|info| {
        if info.callee.span == Some(span) { // Is `span` an expansion root?
            span = self.expansion_root_to_call_site(info.call_site); // note the recursion
        }
    }));
    span
}
```

I think you could also just use `source_callsite`. `souce_callsite(span)` is `expansion_root_to_call_site(span)` iff `expansion_root_to_call_site(span)` has `NO_EXPANSION`, which I believe is the case whenever we report the `unwrap_or` lint -- it appears we already avoid linting expanded code.

With expanded spans, `expansion_root_to_call_site` never makes things worse but does not always lead to a usable snippet, which may not exist.

**Pull request merged**